beacon_node/beacon_chain/src/schema_change/migration_schema_v24.rs

```diff
@@ -0,0 +1,608 @@
+    let mut migrate_ops = vec![];
+    let split = db.get_split_info();
+    let hot_hdiff_start_slot = split.slot;
+
+    // Delete the `PruningCheckpoint` (no longer used).
+    migrate_ops.push(KeyValueStoreOp::DeleteKey(
+        DBColumn::BeaconMeta,
+        PRUNING_CHECKPOINT_KEY.as_slice().to_vec(),
+    ));
+
+    // Sanity check to make sure the HDiff grid is aligned with the epoch start
+    if hot_hdiff_start_slot % T::EthSpec::slots_per_epoch() != 0 {
+        return Err(Error::MigrationError(format!(
+            "hot_hdiff_start_slot is not first slot in epoch {hot_hdiff_start_slot}"
+        )));
+    }
+
+    // After V24 hot tree states, the in-memory `anchor_info.anchor_slot` is the start slot of the
+    // hot HDiff grid. Before the migration, it's set to the slot of the anchor state in the DB:
+    // - the genesis state on a genesis sync, or
+    // - the checkpoint state on a checkpoint sync.
+    //
+    // If the node has been running for a while the `anchor_slot` might be less than the finalized
+    // checkpoint. This upgrade constructs a grid only with unfinalized states, rooted in the
+    // current finalized state. So we set the `anchor_slot` to `split.slot` to root the grid in the
+    // current finalized state. Each migration sets the split to
+    // ```
+    // Split { slot: finalized_state.slot(), state_root: finalized_state_root }
+    // ```
+    {
+        let anchor_info = db.get_anchor_info();
+
+        // If the node is already an archive node, we can set the anchor slot to 0 and copy
+        // snapshots and diffs from the freezer DB to the hot DB in order to establish an initial
+        // hot grid that is aligned/"perfect" (no `start_slot`/`anchor_slot` to worry about).
+        //
+        // This only works if all of the following are true:
+        //
+        // - We have the previous snapshot for the split state stored in the freezer DB, i.e.
+        //   if `previous_snapshot_slot >= state_upper_limit`.
+        // - The split state itself will be stored as a diff or snapshot in the new grid. We choose
+        //   not to support a split state that requires block replay, because computing its previous
+        //   state root from the DAG is not straight-forward.
+        let dummy_start_slot = Slot::new(0);
+        let closest_layer_points = db
+            .hierarchy
+            .closest_layer_points(split.slot, dummy_start_slot);
+
+        let previous_snapshot_slot =
+            closest_layer_points
+                .iter()
+                .copied()
+                .min()
+                .ok_or(Error::MigrationError(
+                    "closest_layer_points must not be empty".to_string(),
+                ))?;
+
+        if previous_snapshot_slot >= anchor_info.state_upper_limit
+            && db
+                .hierarchy
+                .storage_strategy(split.slot, dummy_start_slot)
+                .is_ok_and(|strategy| !strategy.is_replay_from())
```

Comment:
Should maybe test that this doesn't affect replay_from split states during normal operation.

We should add a note about this limitation, in case we delete blocks prior to the split from the hot DB in future.

Reply:
I think it's ok during normal operation because `previous_state_root` is already filled in. The reason we can't migrate a ReplayFrom split state is that we can't compute the previous state root from the DAG (because the split doesn't have any ancestors in the DAG).